### PR TITLE
feat: Create a verifier as a service

### PR DESCRIPTION
**Pull request by laurentsimon**
See https://github.com/slsa-framework/slsa-verifier/issues/163

This is an experimental feature, obviously. This PR does:
1. Re-factor code to be able to share the verification code between the cli and the service
2. Implement a simple REST API `/v1/verify` which takes as input the same as the CLI (except for the artifact which is replaced by its hash):
```yaml
{
    "source": "github.com/ossf/scorecard",
    "artifactHash": "02e7ac9a2a252a4e57cb2050953e787d095a04b955a41878068d60885bd31a0b",
    "tag": "v4.4.0",
     versionedTag: "bla",
    "printProvenance": true,
    "provenanceContent": "..."
}
```

Example: `curl -s 127.0.0.1:8000/v1/verify -d @./request`

and returns:
```yaml
{
   "error": "some caught error"
   "validation": "failure" / "success"
   "provenanceContent": "base64 intoto"
}
```

We may want to change some of the names. I have not included unit tests yet. Will do in a follow-up PR. (need to mock the http client).

Let me know what you think about the input / output format. I'm wondering if we should unify the output for both the cli and the service. It would force us to have proper error / log handling

### CODE REVIEW

File: cmd/verify.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package cmd
```
Comment (laurentsimon) [inline cmd/verify.go:1]: this function was part of main.go, I just moved it in here. No change to the code itself

**Review by naveensrinivasan — COMMENTED** [review]
This is large PR.

Can you please add some explanation as to what are these changes for? This will help others with the review process.

Can you please explain as to why are these test files part of this PR?

**Comment by naveensrinivasan** [comment]
What is the idea behind the service? Is it for SLSA org to host this service? It is not clear

**Comment by laurentsimon** [comment]
> This is large PR.

Only the 3 files under `service/` are new. The rest is renaming files.

> Can you please add some explanation as to what are these changes for? This will help others with the review process.

Please see https://github.com/slsa-framework/slsa-verifier/issues/163

> Can you please explain why are these test files part of this PR?

the folder has been renamed, there is no change to these files.

**Comment by laurentsimon** [comment]
> What is the idea behind the service? Is it for SLSA org to host this service? It is not clear

Please see https://github.com/slsa-framework/slsa-verifier/issues/163

**Comment by naveensrinivasan** [comment]
> > What is the idea behind the service? Is it for SLSA org to host this service? It is not clear
> 
> Please see #163

Thanks for editing the #163 with additional details. Also, a suggestion would be is to include a summary in the PR, which helps anyone new to understand the context before jumping into details about the issue.

File: .github/workflows/pre-submit.yml [hunk 2]
```diff
@@ -20,7 +20,16 @@ jobs:
           set -euo pipefail
 
           go mod vendor
-          # Build
+          
+          # Build cli
+          cd cli
```
Comment (asraa) [inline .github/workflows/pre-submit.yml:25]: simplify to remove the `cd .../` with `go build -mod=vendor ./cli` and `go build -mod=vendor ./service`?

File: service/verify.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+	"github.com/slsa-framework/slsa-verifier/cmd"
+	"github.com/slsa-framework/slsa-verifier/verification"
+)
+
+var errInvalid = errors.New("invalid")
+
+type v1Query struct {
+	// Compulsory fields.
+	Source       string `json:"source"`
+	ArtifactHash string `json:"artifactHash"`
```
Comment (asraa) [inline service/verify.go:21]: for containers, we may want to support this as a oneof artifacthash/dsse envelope or container image reference.

I would rather see grpc service like definitions, but save for a later review. just commenting on the idea for now.

File: service/verify.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+	"github.com/slsa-framework/slsa-verifier/cmd"
+	"github.com/slsa-framework/slsa-verifier/verification"
+)
+
+var errInvalid = errors.New("invalid")
+
+type v1Query struct {
+	// Compulsory fields.
+	Source       string `json:"source"`
```
Comment (asraa) [inline service/verify.go:20]: +1. I think one big mistake in cosign's API was that the source was never pinned. We should never try signature verification without a pinned subject.

File: service/verify.go [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+var (
+	validationSuccess = validation("success")
+	validationFailure = validation("failure")
+)
+
+type v1Result struct {
+	Version         uint       `json:"version"`
+	Error           *string    `json:"error,omitempty"`
+	Validation      validation `json:"validation"`
```
Comment (asraa) [inline service/verify.go:40]: Maybe an exit status int? easier than a string for version compatibility in case we extend this later.

**Review by asraa — COMMENTED** [review]
Overall, I think it's useful to have a docker container service: it would be easy for users to run in their own clusters, re-package with a container image, etc. 

The limitation with signing the result is important too. Users should be able to specify a signer (e.g. we can use  in-memory signers for ephemeral test instances, or GCP signers that can rotate keys).

File: .github/workflows/pre-submit.yml (hunk 2 above)
Comment (laurentsimon) [reply .github/workflows/pre-submit.yml:25]: I tried that but it failed `go: build output "cli" already exists and is a directory`... any way around that?

File: service/verify.go (hunk 3 above)
Comment (laurentsimon) [reply service/verify.go:21]: SG.

File: service/verify.go (hunk 5 above)
Comment (laurentsimon) [reply service/verify.go:40]: string can also be extended, right?
I'm not against using code, just trying to see the additional benefits it gives us.

I'm thinking we may want an `ErrorMsg`, `ErrorCode`... but I'm not sure if Golang handles that easily in practice...unless we wrap all errors with our own structure. Wdut?

**Comment by laurentsimon** [comment]
> The limitation with signing the result is important too. Users should be able to specify a signer (e.g. we can use in-memory signers for ephemeral test instances, or GCP signers that can rotate keys).

+1, we need to take care of this, in particular for a public service.

**Comment by laurentsimon** [comment]
> Overall, I think it's useful to have a docker container service: it would be easy for users to run in their own clusters, re-package with a container image, etc.
> 
> The limitation with signing the result is important too. Users should be able to specify a signer (e.g. we can use in-memory signers for ephemeral test instances, or GCP signers that can rotate keys).

if you have pointers for this, please drop them in. I'd like to take a look, even if it's not part of this PR. Thanks

File: .github/workflows/pre-submit.yml (hunk 2 above)
Comment (asraa) [reply .github/workflows/pre-submit.yml:25]: `-o verifier`? it's weird i can't find any docs on that... i'm pretty sure the behavior for `./cmd/XYZ` is that it writes to XYZ. But maybe not.

Comment (laurentsimon) [reply .github/workflows/pre-submit.yml]: `-o` did the trick. Thanks

**Comment by laurentsimon** [comment]
> Currently, the CLI tool will be go install'd as `cli`. I think you should move it to `./cmd/slsa-verifier` to make it be named as `slsa-verifier`. Service can move under `/experimental/service`. Should the current `./cmd/verify.go` be moved into `verification/` as another module?

that should work. I'll ping once I've made the changes. Thanks

fyi @ianlewis

**Comment by laurentsimon** [comment]
> > Currently, the CLI tool will be go install'd as `cli`. I think you should move it to `./cmd/slsa-verifier` to make it be named as `slsa-verifier`. Service can move under `/experimental/service`. Should the current `./cmd/verify.go` be moved into `verification/` as another module?
> 
> that should work. I'll ping once I've made the changes. Thanks
> 
> fyi @ianlewis

changes made. I've kept `cli` for the code that defines the cli. I've mades all the other changes. PTAL

**Pull request closed without merging**